README.md
# bddtestcoverage

A small package for testing [`testthat`'s BDD functions](https://testthat.r-lib.org/reference/describe.html). 

## Example

```r
install.packages('pak')
library(pak)
pak::pak('mjfrigaard/bddtestcoverage')
# install devtools/usethis
pak::pak('devtools')
devtools::test()
```

## Feature

`Feature`s are testable pieces of functionality.

```gherkin
Feature: tidy_ggp2movies() Function
  As a human
  I want a function that restructures the columns in the ggplot2movies::movies data
  In order to `tidy` the movie data
```

## Scenarios

`Scenarios` are concrete examples illustrating a `Feature`. They include a title, `Given` (the preconditions or initial context for the test), `When` (an action or event), and `Then` (the expected outcome)

```gherkin
  Scenario: Create `genre_count` categorical column 
    Given a `movies_data` dataframe with columns:
    `title`, `year`, `length`, `budget`, `rating`, 
    `votes`, `mpaa`, `Action`, `Animation`, `Comedy`, 
    `Drama`, `Documentary`, `Romance` and `Short`
    When the tidy_ggp2movies() function is applied to the movies_data
    Then the `genre_count` column should be of type integer
    And the new `genre_count` column should represent the sum of genres per movie
    
  Scenario: Create `genre` categorical column 
    Given a `movies_data` dataframe with columns:
    `title`, `year`, `length`, `budget`, `rating`, 
    `votes`, `mpaa`, `Action`, `Animation`, `Comedy`, 
    `Drama`, `Documentary`, `Romance` and `Short`
    When the tidy_ggp2movies() function is applied to the movies_data
    Then the new `genres` column should list the genres per movie separated by commas
    And `genre` should be of type factor
    And `genre` should categorize movies with multiple genres as `Multiple genres`
    
  Scenario: Convert `mpaa` to factor with specified levels
    Given a `movies_data` dataframe with columns:
    `title`, `year`, `length`, `budget`, `rating`, 
    `votes`, `mpaa`, `Action`, `Animation`, `Comedy`, 
    `Drama`, `Documentary`, `Romance` and `Short`
    When the tidy_ggp2movies() function is applied to the movies_data
    Then the `mpaa` column should be of type factor
    And have levels `G`, `PG`, `PG-13`, `R`, `NC-17`

  Scenario: Correct columns in tidy ggplot2movies::movies data
    Given a `movies_data` dataframe with columns:
    `title`, `year`, `length`, `budget`, `rating`, 
    `votes`, `mpaa`, `Action`, `Animation`, `Comedy`, 
    `Drama`, `Documentary`, `Romance` and `Short`
    When the tidy_ggp2movies() function is applied to the movies_data
    Then the output dataframe should have columns:
    `title`,  `year`,  `length`,  `budget`,  `rating`,
    `votes`,  `mpaa`,  `genre_count`,  `genres` and `genre` 
```

## Background

The `Background` keyword can be used to reduce duplication: 

```gherkin
  Background:
    Given a `movies_data` dataframe with columns:
    `title`, `year`, `length`, `budget`, `rating`, 
    `votes`, `mpaa`, `Action`, `Animation`, `Comedy`, 
    `Drama`, `Documentary`, `Romance` and `Short`
    
  Scenario: Create `genre_count` categorical column 
    When the tidy_ggp2movies() function is applied to the movies_data
    Then the `genre_count` column should be of type integer
    And the new `genre_count` column should represent the sum of genres per movie
  
  Scenario: Create `genre` categorical column 
    When the tidy_ggp2movies() function is applied to the movies_data
    Then the new `genres` column should list the genres per movie separated by commas
    And `genre` should be of type factor
    And `genre` should categorize movies with multiple genres as `Multiple genres`
    
  Scenario: Convert `mpaa` to factor with specified levels
    When the tidy_ggp2movies() function is applied to the movies_data
    Then the `mpaa` column should be of type factor
    And have levels `G`, `PG`, `PG-13`, `R`, `NC-17`
    
  Scenario: Correct columns in tidy ggplot2movies::movies data
    When the tidy_ggp2movies() function is applied to the movies_data
    Then the output dataframe should have columns:
    `title`,  `year`,  `length`,  `budget`,  `rating`,
    `votes`,  `mpaa`,  `genre_count`,  `genres` and `genre`
```


NEWS.md


# ISSUES

## BDD `it()` function not loading test helpers

I noticed the new `it()` functions don't seem to load helper functions after `load_all()`. 

### `R/` folder:

```
R
├── bddtestcoverage-package.R
└── collapse_movies_genres.R
```



### `tests/` folder:

```
tests
├── testthat
│   ├── fixtures
│   │   ├── make_test_data.R
│   │   └── test_data.rds
│   ├── helper.R
│   └── test-collapse_movies_genres.R
└── testthat.R
```

### Helper function

In `tests/testthat/helper.R`, I have the following helper:

```r
test_logger <- function(msg, start = NULL, end = NULL) {
  if (is.null(start) & is.null(end)) {
    cat("\n")
    logger::log_info("{msg}")
  } else if (!is.null(start) & is.null(end)) {
    cat("\n")
    logger::log_info("\n[ START {start} = {msg}]")
  } else if (is.null(start) & !is.null(end)) {
    cat("\n")
    logger::log_info("\n[ END {end} = {msg}]")
  } else {
    cat("\n")
    logger::log_info("\n[ START {start} = {msg}]")
    cat("\n")
    logger::log_info("\n[ END {end} = {msg}]")
  }
}
```

In `test-collapse_movies_genres.R`:

When the `test_logger()` helper is inside `it()` (with a nested `describe()`):

```r
testthat::describe(
  description = "
    Feature: collapse_movies_genres Function
      In order to tidy up the movie data
      As a data scientist
      I want to convert ggplot2moves::movies into a 'tidy' data.frame", code = {
    testthat::describe(
      description = "
      Background:
      Given a dataframe 'movies_data' with columns:
        'title', 'year', 'length', 'budget', 'rating',
        'votes', 'mpaa', 'Action', 'Animation', 'Comedy',
        'Drama', 'Documentary', 'Romance', and 'Short'", code = {
        testthat::it(
          description = "
      Scenario: Correct formatting of mpaa as factor with specified levels and labels
          When the function collapse_movies_genres is applied to the movies_data
          Then the 'mpaa' column should be of type factor
          And have levels 'G', 'PG', 'PG-13', 'R', 'NC-17'",
          code = {
            test_logger(start = "fixture", msg = "test_data.rds")
            # load fixture
            test_data <- readRDS(file = test_path("fixtures", "test_data.rds"))
            tidy_ggp2 <- collapse_movies_genres(test_data)
            testthat::expect_true(is.factor(tidy_ggp2$mpaa))
            testthat::expect_equal(
              object = levels(tidy_ggp2$mpaa),
              expected = c("G", "PG", "PG-13", "R", "NC-17")
            )
            test_logger(end = "fixture", msg = "test_data.rds")
          })
      })
  })

```

```r
devtools::load_all(".")
ℹ Loading bddtestcoverage
```

The test fails (can't find the helper).

```r
devtools:::test_active_file()
[ FAIL 1 | WARN 0 | SKIP 0 | PASS 0 ]

── Error (test-collapse_movies_genres.R:22:13): 
      Scenario: Correct formatting of mpaa as factor with specified levels and labels
          When the function collapse_movies_genres is applied to the movies_data
          Then the 'mpaa' column should be of type factor
          And have levels 'G', 'PG', 'PG-13', 'R', 'NC-17' ──
Error in `test_logger(start = "fixture", msg = "test_data.rds")`: could not find function "test_logger"
[ FAIL 1 | WARN 0 | SKIP 0 | PASS 0 ]
```

And if I switch it to a single `describe()` function: 

```r
testthat::describe(
  description = "
    Feature: collapse_movies_genres Function
      In order to tidy up the movie data
      As a data scientist
      I want to convert ggplot2moves::movies into a 'tidy' data.frame", code = {
    testthat::it(
      description = "
      Scenario: Correct formatting of mpaa as factor with specified levels and labels
        Given a dataframe 'movies_data' with columns:
          'title', 'year', 'length', 'budget', 'rating',
          'votes', 'mpaa', 'Action', 'Animation', 'Comedy',
          'Drama', 'Documentary', 'Romance', and 'Short'
        When the function collapse_movies_genres is applied to the movies_data
        Then the 'mpaa' column should be of type factor
        And have levels 'G', 'PG', 'PG-13', 'R', 'NC-17'",
      code = {
        test_logger(start = "fixture", msg = "test_data.rds")
        # load fixture
        test_data <-
          readRDS(file = test_path("fixtures", "test_data.rds"))
        tidy_ggp2 <- collapse_movies_genres(test_data)
        testthat::expect_true(is.factor(tidy_ggp2$mpaa))
        testthat::expect_equal(
          object = levels(tidy_ggp2$mpaa),
          expected = c("G", "PG", "PG-13", "R", "NC-17")
        )
        test_logger(end = "fixture", msg = "test_data.rds")
      })
  })
```

```r
devtools::load_all(".")
ℹ Loading bddtestcoverage
```

The test still fails:

```r
[ FAIL 1 | WARN 0 | SKIP 0 | PASS 0 ]

── Error (test-collapse_movies_genres.R:54:9): 
      Scenario: Correct formatting of mpaa as factor with specified levels and labels
        Given a dataframe 'movies_data' with columns:
          'title', 'year', 'length', 'budget', 'rating',
          'votes', 'mpaa', 'Action', 'Animation', 'Comedy',
          'Drama', 'Documentary', 'Romance', and 'Short'
        When the function collapse_movies_genres is applied to the movies_data
        Then the 'mpaa' column should be of type factor
        And have levels 'G', 'PG', 'PG-13', 'R', 'NC-17' ──
Error in `test_logger(start = "fixture", msg = "test_data.rds")`: could not find function "test_logger"
[ FAIL 1 | WARN 0 | SKIP 0 | PASS 0 ]
```

But if the helper is inside `test_that()` (with nested `describe()`):

```r
testthat::describe(
  description = "
    Feature: collapse_movies_genres Function
      In order to tidy up the movie data
      As a data scientist
      I want to convert ggplot2moves::movies into a 'tidy' data.frame", code = {
    testthat::describe(
      description = "
      Background:
      Given a dataframe 'movies_data' with columns:
        'title', 'year', 'length', 'budget', 'rating',
        'votes', 'mpaa', 'Action', 'Animation', 'Comedy',

tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
# * https://testthat.r-lib.org/articles/special-files.html

library(testthat)
library(testcoverage)

test_check("testcoverage")


tests/testthat/helper.R
test_logger <- function(msg, start = NULL, end = NULL) {
  if (is.null(start) & is.null(end)) {
    cat("\n")
    logger::log_info("{msg}")
  } else if (!is.null(start) & is.null(end)) {
    cat("\n")
    logger::log_info("\n[ START {start} = {msg}]")
  } else if (is.null(start) & !is.null(end)) {
    cat("\n")
    logger::log_info("\n[ END {end} = {msg}]")
  } else {
    cat("\n")
    logger::log_info("\n[ START {start} = {msg}]")
    cat("\n")
    logger::log_info("\n[ END {end} = {msg}]")
  }
}


tests/testthat/test-tidy_ggp2movies.R

# FIXTURE DATA TESTS: ----
# Thet tests below use the data from tests/testthat/fixtures/test_data.rds
# helper called nside it() (in two nested describe() calls)  ==================
describe("Feature: (As a ..., In order to ..., I want to ...)", {
  describe("Background (Given ...)", {
      it("Scenario: (Given ... , When ..., Then ...)", {
          ## call helper ----
          test_logger(start = "mpaa", msg = "is.factor()/levels()")
          # load fixture
          test_data <- readRDS(file = test_path("fixtures", "test_data.rds"))
          tidy_ggp2 <- tidy_ggp2movies(test_data)
          expect_true(is.factor(tidy_ggp2$mpaa))
          expect_equal(
            object = levels(tidy_ggp2$mpaa),
            expected = c("G", "PG", "PG-13", "R", "NC-17")
          )
          test_logger(end = "mpaa", msg = "is.factor()/levels()")
        })
    })
})

# helper called inside test_that() (in two nested describe() calls) ============
describe("Feature: (As a ..., In order to ..., I want to ...)", {
    describe("Background (Given ...)", {
        test_that("Scenario: (Given ... , When ..., Then ...)", {
            ## call helper ----
            test_logger(start = "mpaa", msg = "is.factor()/levels()")
            # load fixture
            test_data <- readRDS(file = test_path("fixtures", "test_data.rds"))
            tidy_ggp2 <- tidy_ggp2movies(test_data)
            expect_true(is.factor(tidy_ggp2$mpaa))
            expect_equal(
              object = levels(tidy_ggp2$mpaa),
              expected = c("G", "PG", "PG-13", "R", "NC-17")
            )
            test_logger(end = "mpaa", msg = "is.factor()/levels()")
          })
      })
  })

# helper called inside it() with single describe() ============================
describe("Feature: (As a ..., In order to ..., I want to ...)", {
    it("Scenario: (Given ... , When ..., Then ...)", {
        ## call helper ----
        test_logger(start = "mpaa", msg = "is.factor()/levels()")
        # load fixture
        test_data <- readRDS(file = test_path("fixtures", "test_data.rds"))
        tidy_ggp2 <- tidy_ggp2movies(test_data)
        expect_true(is.factor(tidy_ggp2$mpaa))
        expect_equal(
          object = levels(tidy_ggp2$mpaa),
          expected = c("G", "PG", "PG-13", "R", "NC-17")
        )
        test_logger(end = "mpaa", msg = "is.factor()/levels()")
      })
  })

# nested describe()s, helper called inside test_that()  =================
describe("Feature: (As a ..., In order to ..., I want to ...)", {
    describe("Background: (Given ...)", {
        test_that("Scenario: (Given ... , When ..., Then ...)", {
            ## call helper ----
            test_logger(start = "mpaa", msg = "is.factor()/levels()")
            ## load fixture ----
            test_data <- readRDS(file = test_path("fixtures", "test_data.rds"))

            tidy_ggp2 <- tidy_ggp2movies(test_data)
            expect_true(is.factor(tidy_ggp2$mpaa))
            expect_equal(
              object = levels(tidy_ggp2$mpaa),
              expected = c("G", "PG", "PG-13", "R", "NC-17")
            )
            test_logger(end = "mpaa", msg = "is.factor()/levels()")
          })
      })
  })

# helper used outside it(), but inside nested describe() =====================
describe("Feature: (As a ..., In order to ..., I want to ...)", {
    describe("Background: (Given ...)", {
        ## call helper ----
        test_logger(start = "mpaa", msg = "is.factor()/levels()")
        it("Scenario: (Given ... , When ..., Then ...)", {
            # load fixture
            test_data <- readRDS(file = test_path("fixtures", "test_data.rds"))
            tidy_ggp2 <- tidy_ggp2movies(test_data)
            expect_true(is.factor(tidy_ggp2$mpaa))
            expect_equal(
              object = levels(tidy_ggp2$mpaa),
              expected = c("G", "PG", "PG-13", "R", "NC-17")
            )
          })
        test_logger(end = "mpaa", msg = "is.factor()/levels()")
      })
  })


R/tidy_ggp2movies.R
#' Test function (tidy ggplot2movies::movies)
#'
#' @param movies_data `ggplot2movies::movies` data
#'
#' @return A tidy `movies` `data.frame`
#'
#' @export tidy_ggp2movies
#'
#'
#' @examples
#' require(ggplot2movies)
#' tidy_ggp2movies(ggplot2movies::movies)
tidy_ggp2movies <- function(movies_data) {
  # specify genre columns
  genre_cols <- c(
    "Action", "Animation",
    "Comedy", "Drama",
    "Documentary", "Romance",
    "Short"
  )
  # calculate row sum for genres
  movies_data$genre_count <- rowSums(movies_data[, genre_cols])
  # create aggregate 'genres' for multiple categories
  movies_data$genres <- apply(
    X = movies_data[, genre_cols],
    MARGIN = 1,
    FUN = function(row) {
      genres <- names(row[row == 1])
      if (length(genres) > 0) {
        return(paste(genres, collapse = ", "))
      } else {
        return(NA)
      }
    }
  )
  # format variables
  movies_data$genre_count <- as.integer(movies_data$genre_count)
  movies_data$genre <- ifelse(test = movies_data$genre_count > 1,
    yes = "Multiple genres",
    no = movies_data$genres
  )
  movies_data$genre <- as.factor(movies_data$genre)
  movies_data$mpaa <- factor(movies_data$mpaa,
    levels = c("G", "PG", "PG-13", "R", "NC-17"),
    labels = c("G", "PG", "PG-13", "R", "NC-17")
  )

  # reduce columns to only those in graph
  movies_data[, c(
    "title", "year", "length", "budget",
    "rating", "votes", "mpaa", "genre_count",
    "genres", "genre"
  )]
}


R/collapse_movies_genres.R
#' Test function (tidy ggplot2movies::movies)
#'
#' @param movies_data `ggplot2movies::movies` data
#'
#' @return A tidy `movies` `data.frame`
#'
#' @export collapse_movies_genres
#'
#'
#' @examples
#' require(ggplot2movies)
#' collapse_movies_genres(ggplot2movies::movies)
collapse_movies_genres <- function(movies_data) {
  # specify genre columns
  genre_cols <- c(
    "Action", "Animation",
    "Comedy", "Drama",
    "Documentary", "Romance",
    "Short"
  )
  # calculate row sum for genres
  movies_data$genre_count <- rowSums(movies_data[, genre_cols])
  # create aggregate 'genres' for multiple categories
  movies_data$genres <- apply(
    X = movies_data[, genre_cols],
    MARGIN = 1,
    FUN = function(row) {
      genres <- names(row[row == 1])
      if (length(genres) > 0) {
        return(paste(genres, collapse = ", "))
      } else {
        return(NA)
      }
    }
  )
  # format variables
  movies_data$genre_count <- as.integer(movies_data$genre_count)
  movies_data$genre <- ifelse(test = movies_data$genre_count > 1,
    yes = "Multiple genres",
    no = movies_data$genres
  )
  movies_data$genre <- as.factor(movies_data$genre)
  movies_data$mpaa <- factor(movies_data$mpaa,
    levels = c("G", "PG", "PG-13", "R", "NC-17"),
    labels = c("G", "PG", "PG-13", "R", "NC-17")
  )

  # reduce columns to only those in graph
  movies_data[, c(
    "title", "year", "length", "budget",
    "rating", "votes", "mpaa", "genre_count",
    "genres", "genre"
  )]
}
